### PR TITLE
fix loss with CP enabled

### PR DESCRIPTION
**Pull request by tianyu-l**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #761

This PR
1. fixes the loss all-reduce mesh from `dp_replicate + dp_shard`, to `dp_replicate + dp_shard + cp`, by creating a new submesh named by `"dp_cp"`
2. removes the restriction on running `dp_replicate` + `cp` with `dp_shard=1`, as it's just HSDP with the FSDP part being CP only; also added a test of this case
3. fixes the estimation.py parallelism condition checking, as it only works with HSDP and FSDP

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -336,10 +336,14 @@ def loss_fn(pred, labels):
...
-                if parallel_dims.dp_enabled:
+                if (
+                    parallel_dims.dp_replicate_enabled
+                    or parallel_dims.dp_shard_enabled
+                    or parallel_dims.cp_enabled
+                ):
                     global_avg_loss, global_max_loss = (
-                        utils.dist_mean(avg_loss, dp_mesh),
-                        utils.dist_max(max_loss, dp_mesh),
+                        utils.dist_mean(avg_loss, world_mesh["dp_cp"]),
```
Comment (awgu) [inline train.py:345]: something to consider for the future:
constructing NCCL communicators may use extra GPU memory
depending on the tradeoff, doing 2 separate all-reduces over existing communicators may some memory over creating a new communicator to do a single all-reduce

**Review by awgu — APPROVED** [review]

File: train.py (hunk 1 above)
Comment (tianyu-l) [reply train.py:345]: Makes sense. Tbh I'd hope device mesh supports all-reduce over a 2-D mesh and handles that for me, but maybe this use case is not be popular enough.

**Comment by fegin** [comment]
Good catch about the loss issue. As for 2-D mesh allreduce, I think it's more about the functional collective supporting this or not. DeviceMesh is just a representation of device mesh.
